subsys/bluetooth/audio/has.c

```diff
@@ -34,11 +34,96 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
 				 sizeof(has.features));
 }
 
+#if CONFIG_BT_HAS_PRESET_COUNT > 0
+static uint8_t handle_control_point(struct bt_conn *conn, uint8_t opcode,
+				    struct net_buf_simple *buf)
+{
+	BT_DBG("conn %p opcode %s (0x%02x)", conn, bt_has_op_str(opcode), opcode);
+
+	/* TODO: handle request here */
+
+	return BT_HAS_ERR_INVALID_OPCODE;
+}
+
+static ssize_t write_control_point(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				   const void *data, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	const struct bt_has_cp_hdr *hdr;
+	struct net_buf_simple buf;
+	uint8_t err;
+
+	BT_DBG("conn %p attr %p data %p len %d offset %d flags 0x%02x", conn, attr, data, len,
+	       offset, flags);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len < sizeof(*hdr)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, len);
+
+	hdr = net_buf_simple_pull_mem(&buf, sizeof(*hdr));
+
+	err = handle_control_point(conn, hdr->opcode, &buf);
+
+	return (err != BT_ATT_ERR_SUCCESS) ? BT_GATT_ERR(err) : len;
+}
+
+static ssize_t read_active_preset_index(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					void *buf, uint16_t len, uint16_t offset)
+{
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > sizeof(has.active_index)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &has.active_index,
+				 sizeof(has.active_index));
+}
+
+static ssize_t cp_ccc_cfg_write(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				uint16_t value)
+{
+	BT_DBG("conn %p, value 0x%04x", conn, value);
+
+	/* TODO: handle subscription */
+
+	return sizeof(value);
+}
+
+static void ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	BT_DBG("attr %p value 0x%04x", attr, value);
+}
+
+/* Preset Control Point CCC */
+static struct _bt_gatt_ccc cp_ccc_cfg =
+	BT_GATT_CCC_INITIALIZER(ccc_cfg_changed, cp_ccc_cfg_write, NULL);
```

Comment:
Do we need to do anything in the `cp_ccc_cfg_write`? 

Otherwise suggest to use the `MANAGED` CCC version

Reply:
I can use the `MANAGED` version for now. But I have plans to make use of `value` to know whether client subscribed for notifications or indications or both.
The `_changed` callback from `MANAGED` CCC version does not give such information.